Doxygen documentation:  Request for development community comments

MatthewMasarik-NOAA opened the pull request:
# Pull Request Summary                                                                                                                                                                                                                                                                                                                                      
Templates for _doxygen_ documentation.                                                                                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                                                                                            
## Description                                                                                                                                                                                                                                                                                                                                              
This PR seeks adoption of a set of templates for each Fortran 'code unit'  (file, program, subroutine, etc).  Each code unit has a section below with the suggested _required_ and _optional_ tags for the template.  Links to samples of _input_ and _output_ for the templates are given for each as well.  At this time only acceptance of the templates is being considered -- the marked up code and output will follow in a separate PR.

<br>

**UPDATED:  Jan 10th - Addresses comments made by** @edwardhartnett

<br>

### Notes
*  The complete doxygen documentation is hosted at:  [https://matthewmasarik-noaa.github.io/WW3](https://matthewmasarik-noaa.github.io/WW3).
* For reference, the complete list of Doxygen [Special Commands (Tags)](https://www.doxygen.nl/manual/commands.html) _Note: the commands are shown using the `\` version, these are equivalent to the `@` version._ 
* Note below:  _doxygen determines the end of the `@brief` section by the first period (keep in mind when using '.' in punctuation before the end of the sentence).  The blank line following the `@brief` sentence is not required, but is customary._ 

<br>

### Templates
#### __FILE__
  * _Required_
  ```
  !> @file  <file name>
  !> @brief  <one sentence description>
  !> 
  !> @author <author name> @date <dd-Mon-yyyy>
  ``` 
  * _Optional_:  `@todo`, `@remark`,`@remark`,`@attention`,`@warning`, `@copyright`.
  * _Sample_ 
    * Input:  [w3updtmd.F90](https://github.com/MatthewMasarik-NOAA/WW3/blob/dx1/model/src/w3updtmd.F90#L1-L4),
    * Output:  [w3updtmd.F90](https://matthewmasarik-noaa.github.io/WW3/w3updtmd_8F90.html).
                                                                                                                                                                                                                                                                                                                  
<br> 

#### __MODULE__
  * _Required_
```
!> @brief <one sentence description>    # contents of _1. Purpose_ from original header
!> 
!> @details <extended description>
!>
!> @param <module var 1>
!> ...
!> @param <module var N>
!> @author <author name> @date <dd-Mon-yyyy>
```
  * _Note_:  ALL module variables need documentation. 
  * _Optional_:  `@copyright`,`@cite`,`@example`,`@code`,`@todo`, `@bug`,`@remark`,`@attention`,`@warning`.
  * _Sample_ 
    * Input: [w3swldmd](https://github.com/MatthewMasarik-NOAA/WW3/blob/dx1/model/src/w3swldmd.F90#L10-L16), 
    * Output:  [w3swldmd](https://matthewmasarik-noaa.github.io/WW3/namespacew3swldmd.html).

<br>

####  __PROGRAM__
   * _Required_
```
!> @brief  <one sentence description>      # contents of _1. Purpose_ from original header
!>
!> @details <extended description>          # contents of _2. Method_ from original header
!> 
!> @author <author name>  @date <dd-Mon-yyyy>
```
   * _Optional_: `@copyright`,`@cite`,`@example`,`@code`,`@todo`, `@bug`,`@remark`,`@attention`,`@warning`.
  * _Sample_
    * Input: [w3grib](https://github.com/MatthewMasarik-NOAA/WW3/blob/dx1/model/src/ww3_grib.F90#L12-L27), 
    * Output:  [w3grib](https://matthewmasarik-noaa.github.io/WW3/ww3__grib_8F90.html#a7ba9223f46f879df219348b44809abe1).

<br>

#### __SUBROUTINE__
  * _Required_
```
!> @brief  <one sentence description>       # contents of _1. Purpose_ from original header
!>
!> @details <extended description>           # contents of _2. Method_ from original header
!>
!> @param[in/out] <param name> <very short description> # specify:  [in], [out], [inout], or just `@param`.
!> @author <author name>  @date <dd-Mon-yyyy>
!>
```
  * _Optional_:  `@copyright`,`@cite`,`@example`,`@code`,`@todo`, `@bug`,`@remark`,`@attention`,`@warning`.
  * _Sample_  
    * Input: [w3dzxy()](https://github.com/MatthewMasarik-NOAA/WW3/blob/dx1/model/src/w3updtmd.F90#L3137-L3146), 
    * Output:  [w3dzxy()](https://matthewmasarik-noaa.github.io/WW3/namespacew3updtmd.html#af512f4f684985cd9e281d986e72a5d57).

<br>

#### __FUNCTION__
  * _Required_
```
!> @brief  <one sentence description>
!> @param <param name>
!> @returns <return variable name>
```
  * _Optional_:  `@author`, `@date`, `@details`, `@cite`,`@example`,`@code`,`@todo`, `@bug`,`@remark`,`@attention`,`@warning`.
  * _Sample_  
    * Input: [juldayt()](https://github.com/MatthewMasarik-NOAA/WW3/blob/dx1/model/src/w3tidemd.F90#L1261-L1268), 
    * Output: [juldayt()](https://matthewmasarik-noaa.github.io/WW3/namespacew3tidemd.html#ad251f05352f3caf1586b9939e38525ec).

<br>

#### __GLOBAL VARIABLES__ (inline documentation)
  * _Required_
```
!<  <inline comment>
```
  * _Optional_: N/A
  * _Sample_  
    * Input:  [ptmeth](https://github.com/MatthewMasarik-NOAA/WW3/blob/dx1/model/src/w3odatmd.F90#L441), 
    * Output:  [w3odatmd.F90](https://matthewmasarik-noaa.github.io/WW3/w3odatmd_8F90.html) see under section _Variables_: noge, idout, ptmeth, ptfcut.

<br>
                                                                                                                                                                                                                                                                                                      
Please also include the following information:                                                                                                                                                                                                                                                                                                              
* Mention any labels that should be added:                                                                                                                                                                                                                                                                                                                  
  * _documentation_                                                                                                                                                                                                                                                                                                                                         
* Are answer changes expected from this PR?                                                                                                                                                                                                                                                                                                                 
  * No.  There are no changes to the repository at this point, only the acceptance of the templates is being considered.                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                                                                            
### Issue(s) addressed                                                                                                                                                                                                                                                                                                                                      
* Please list any issues associated with this PR, including those the PR will fix/close.                                                                                                                                                                                                                                                                    
  * This works towards addressing noaa-emc/ww3/issues/[#547](https://github.com/NOAA-EMC/WW3/issues/547), though it will not close it.  A separate PR with the actual markup to the source code will follow.                                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                                                                                            
### Commit Message                                                                                                                                                                                                                                                                                                                                          
Accept doxygen templates.                                                                                                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                                                                                            
### Check list                                                                                                                                                                                                                                                                                                                                              
- [x] Branch is up to date with the authoritative repository (NOAA-EMC) develop branch.                                                                                                                                                                                                                                                                     
- [x] Checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop).                                                                                                                                                                                  
- [na] If a version number update is required, checked the [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) checklist.                                                                                                                                                                  
- [na] If a new feature was added, a regression test for testing the new feature is added.                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                                                                                            
### Testing                                                                                                                                                                                                                                                                                                                                                 
* How were these changes tested?                                                                                                                                                                                                                                                                                                                            
  * N/A.  Only the templates are being formally accepted -- no changes to the repo at this time.                                                                                                                                                                                                                                                                                                                           
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)                                                                                                                                                                                                                                                                
  * N/A.                                                                                                                                                                                                                                                                        
* Have the matrix regression tests been run (if yes, please note HPC and compiler)?                                                                                                                                                                                                                                                                         
  * N/A.                                                                                                                                                                                                                                                                                                                             
* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_):                                                                                                                                                                                                                                   
  * N/A.                                                                                                                                                                                                                                                                                                      
* Please indicate the expected changes in the regression test output ([Note the known list of non-identical tests](https://github.com/NOAA-EMC/WW3/wiki/How-to-use-matrix.comp-to-compare-regtests-with-master#4-look-at-results)).                                                                                                                         
  * N/A.                                                                                                                                                                                                                                                                                                                                        
* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_):                                                                                                                                                                                                                                   
  * N/A.

MatthewMasarik-NOAA committed:
update doxygen

MatthewMasarik-NOAA committed:
Doxyfile: add

JessicaMeixner-NOAA commented on model/src/w3odatmd.F90:
```diff
@@ -1,5 +1,20 @@
+!
+!> @author H. L. Tolman @date 22-Mar-2021
+!>
+!> @brief Define data structures to set up wave model grids and aliases
```
brief is included twice? 

Also if the #include statement is moved to after all doxygen comments does it disappear from the doxygen documentation?

JessicaMeixner-NOAA replied:
oh or is one brief for the whole file and the other is for the module?

MatthewMasarik-NOAA replied:
Yes, to your second comment.  I did a brief for the file, and one for the module.  In these cases we might put the same, or similar, descriptions.  Since each is a separate component, it will be used for it's respective display.

MatthewMasarik-NOAA replied:
We discussed this offline, but for the rest of the community, unfortunately I could not find a way to get rid of the `#include w3macros.h` statement at the top.

JessicaMeixner-NOAA replied:
Thanks @MatthewMasarik-NOAA I"m going to resolve this.

MatthewMasarik-NOAA commented:
Hello all, please review and make any suggestions or comments.   We are planning for this to be open for community comments up until Friday, Jan 14th.

edwardhartnett commented on model/src/w3odatmd.F90:
```diff
@@ -1,5 +1,20 @@
+!> @file w3odatmd.F90
+!> @ brief Define data structures to set up wave model grids and aliases
+!>         to use individual grids transparently.
```
In general, it's not worth trying to maintain any special spacing with doxygen comments. Doxygen formats everything nicely, ignoring your whitespace, and it's just extra work for the programmers who have to maintain the code.

So our documentation blocks on NCEPLIBS look like this:

```
!> @file
!> @brief Create new GRIB2 file with fields from an existing GRIB2 file.
!> @author Stephen Gilbert @date 2004-05-17

!> Program reads selected GRIB2 fields from a file, adds a TOC Flag
!> Field separator block and WMO Header in front of each GRIB2 field,
!> and writes them out to a new file. The output file is in the format
!> required for TOC's FTP Input Service, which can be used to
!> disseminate the GRIB2 bulletins.
!>
!> Passing filenames into this program is done in a special
!> way. Before running this program, set the following environment
!> variables:
!>
!> Variable | Meaning
!> ---------|--------
!> FORT11 | Name of the input GRIB2 file.
!> FORT31 | Name of the input GRIB2 index file.
!> FORT51 | Name to use for the output GRIB2 file.
!>
!> @note The "EXTRACT" variable in the namelist allows users to choose
!> whether they want the entire GRIB2 message containing the requested
!> field (extract=.false.), or a GRIB2 message containing only the
!> requested field (extract=.true.). Both options return the same
!> message if the requested field is the only field in the GRIB2
!> message.
!>
!> ### Program History Log
!> Date | Programmer | Comments
!> -----|------------|---------
!> 2004-05-17 | Gilbert | Initial
!> 2012-06-14 | Vuong | Modified write statement with format specifier
!> 2012-10-22 | Vuong | Changed variable envvar to character*6
!> 2016-10-15 | Vuong | Increased length of file name to 200 Characters

```

(Note that this uses markdown to add a table to the docs. But in all cases, just one space is used as whitespace.)

MatthewMasarik-NOAA replied:
That's great you can include markdown tables like that.  I did discover doxygen will ignore your spacing attempts, this was accidentally left in.  I'll change it.

edwardhartnett commented on model/src/w3odatmd.F90:
```diff
@@ -323,10 +338,10 @@ MODULE W3ODATMD
       INTEGER, PARAMETER      :: NOGRP = 10
```
All module variables need documentation.

If not, then each undocumented variable will generate a documentation warning. The goal is to have warning-free documentation builds, so that new PRs can be checked for documentation completeness. If the code is already full of warnings, that won't work.

So all module variables should be documentat, as NOGE and IDOUT are below.

MatthewMasarik-NOAA replied:
I was not aware of this.  I'll add this to the module template.

edwardhartnett commented on model/src/w3swldmd.F90:
```diff
@@ -1,5 +1,16 @@
+!> @author H. L. Tolman @date 21-Nov-2011
+!> @brief Source term module for swell dissipation.
+!>
+!> @details swell dissipation based on different
```
Proper capitalization of sentences allows readers to easily see where one sentence ends, and another begins. Additionally, in general, we can assume some readers will have English as a second language. Proper capitalization helps them understand your documentation.

MatthewMasarik-NOAA replied:
Yes, I will edit this.

edwardhartnett commented on model/src/w3swldmd.F90:
```diff
@@ -198,6 +222,26 @@ SUBROUTINE W3SWL4 (A, CG, WN, DAIR, S, D)
       END SUBROUTINE W3SWL4
 !/ ------------------------------------------------------------------- /
+!
+!> @author H. L. Tolman @date 16-Feb-2012
```
We usually put the brief description at the top, and the author/date at the bottom of the documentation block.

MatthewMasarik-NOAA replied:
Sounds good, we can follow that convention.  I'll make the change.

edwardhartnett commented on model/src/w3swldmd.F90:
```diff
@@ -198,6 +222,26 @@ SUBROUTINE W3SWL4 (A, CG, WN, DAIR, S, D)
+!> @returns S,D        
+!> @brief Turbulent dissipation of narrow-banded swell
+!> 
+!> @details
```
The details tag is not needed. Nor is the brief tag for subroutines/functions. In the case of a subroutine/function, doxygen will assume the first sentence is the brief description, and the rest is the detailed description, which is a very natural default.

MatthewMasarik-NOAA replied:
Good to know.  So, the brief tag is needed for things other than subroutines/functions, and the details tag is never needed?

MatthewMasarik-NOAA replied:
For the time being I decided to leave the brief and details tags in.  My logic is based on a couple considerations.  The first being that it adds some organization for programmers looking at the code.  It also adds consistency to the headers -- so you don't need to remember which does or doesn't need the tags.  I've added a note to both the subroutine and function templates above about this. I'll be fine either way, so I'm looking forward what other's prefer.

edwardhartnett commented on model/src/w3swldmd.F90:
```diff
@@ -362,7 +406,20 @@ SUBROUTINE W3SWL6 (A, CG, WN, S, D)
 !/ ------------------------------------------------------------------- /
-!/
+!
+!> @private
```
Probably you don't want to use the private tag. We have the EXTRACT_PRIVATE setting on, and, in general for science codes, we want documentation for all functions, even functions with private fortran scope.

MatthewMasarik-NOAA replied:
OK, I agree.  Will make this change.

edwardhartnett commented on model/src/w3tidemd.F90:
```diff
@@ -1,3 +1,6 @@
+!> @file w3tidemd.F90
+!> @brief Contains MODULE W3TIDEMD for tidal analysis of time series.
```
We put an author at the file level in all cases, as well as an author for all functions and subroutines. Sometimes this seems redundant, but they are different things, and doxygen shows them in different places...

MatthewMasarik-NOAA replied:
I know exactly what you're saying.  Good point, I'll make the change.

edwardhartnett commented on model/src/w3updtmd.F90:
```diff
@@ -974,6 +1009,17 @@ SUBROUTINE W3UTAU ( FLFRST )
       END SUBROUTINE W3UTAU
 !/ ------------------------------------------------------------------- /
+!
+!> @author H. L. Tolman @date 06-Jun-2018
```
If you consider how we want this documentation to look, you can use the same order in the comments. For example:

https://noaa-emc.github.io/NCEPLIBS-g2c/g2__create_8c.html#a0e8970973e1cefc142240f961a796108

You will see that the order is:

brief description
detailed descrion
parameters
author/date

So the comment below becomes (note that, in this order, the brief and details tags are assumed by doxygen to be the first sentence, and the rest of the text before the parameters).

!> Initialize the wave field with fetch-limited spectra before the
!> actual calculation start.
!> 
!> (Named as an update routine due to placement in code.) Fetch-limited
!> JONSWAP spectra with a cosine^2 directional distribution and a mean direction
!> taken from the wind.
!>
!> @param[out] A Action density spectra
!> @returns A Action density spectra
!> @author H. L. Tolman @date 06-Jun-2018

MatthewMasarik-NOAA replied:
Interesting, I didn't know about that behavior.  I think that's an efficient convention to follow.  I'll make that change.

MatthewMasarik-NOAA replied:
The ordering changes have been made for all the header templates.  As I mention in a comment above I've left the brief and details tags in currently, though have added a note so others are aware they are not necessary.

edwardhartnett commented on model/src/w3updtmd.F90:
```diff
@@ -1432,6 +1486,15 @@ SUBROUTINE W3UBPT
+!
+!> @author C. Sevigny  @date 27-Aug-2015
+!> @param[in] FLFRST Flag for first pass through routine.
+!> @returns N/A
```
Don't put a return tag for subroutines.

MatthewMasarik-NOAA replied:
I agree.  I'll make this change.

edwardhartnett commented on model/src/w3updtmd.F90:
```diff
@@ -1648,7 +1723,24 @@ SUBROUTINE W3UIC5( FLFRST )
+!> @brief Update ice map in the wave model.
+!> 
+!> @details
+!> @parblock
```
Don't use the parblock tag. Instead, just leave a blank line between two paragraphs, as you are already doing. So:

```
!> The first sentence is the brief description.
!>
!> Now we have another paragraph of information.
!>
!> Now we have yet another paragraph.

```

MatthewMasarik-NOAA replied:
That's good to know, I had thought the parblock was necessary.  I'll make this change.

MatthewMasarik-NOAA replied:
All parblocks have been removed.

edwardhartnett commented on model/src/w3updtmd.F90:
```diff
@@ -3004,6 +3130,18 @@ SUBROUTINE W3UTRN ( TRNX, TRNY )
+!> @param[in] ZUNIT Units of ZZ (used for test output).
+!> @param[out] DZZDX Derivative in X-direction (W-E).
+!> @param[out] DZZDY Derivative in Y-direction (S-N).
+!> @returns DZZDX, DZZDY
```
Don't use the return tag to describe the [out] parameters. A subroutine does not return anything. You have already correctly documented the two parameters as [out] and that's all you need to do.

So just remove the return tag from all subroutine documentation blocks.

MatthewMasarik-NOAA replied:
That's a good point.  I'll make that change.

MatthewMasarik-NOAA replied:
I'll leave this (and others) unresolved momentarily, then will resolve them all at once when I've re-ran doxygen and reposted the output.

edwardhartnett reviewed (changes requested):
See in-line comments...

edwardhartnett commented:
OK, I just noticed that you have a text description of your template for documentation. My comments to the code should also be applied to the template...

edwardhartnett commented:
OK, I see you have this up on your own website. But GitHub provides a way to host your built documentation, called the gh-pages branch.

Here's how to do it:

1. build a copy of the documentation in your clone of the repo.
2. clone a new copy of the repo (an extra clone).
3. create a branch (in the extra clone) called 'gh-pages' and checkout gh-pages.
4. git rm -rf * (removes everything in the gh-pages branch)
5. copy the built documentation to the gh-pages directory: cp <WW3>/build/html/* . (Where <WW3> is the location of your first clone, where you built the docs.)
6. git add *
7. git commit -m "updated docs" -a && git push

Now your documentation will appear at: https://noaa-emc.github.io/WW3/. Put a link in your README.md. (See https://github.com/NOAA-EMC/NCEPLIBS-g2 for example.)

(The gh-pages feature also has to be turned on in the project settings for the repo.)

MatthewMasarik-NOAA committed:
updates for Eds
 comments

edwardhartnett reviewed (approved):
Looking real good! Do you want some help setting up the Doxyfile.in?

Also how do you handle version numbers? In the NCEPLIBS we have a file at the top level called VERSION which contains the version number. From there it is pulled in by the build system into the docs and the code.

MatthewMasarik-NOAA commented:
> Looking real good! Do you want some help setting up the Doxyfile.in?
> 

Sure.  If you guys have a way you tend to set them up I'm interested to hear.  The file I've been working with is at:   [Doxyfile](https://github.com/MatthewMasarik-NOAA/WW3/blob/dx1/Doxyfile).  I've made a few small recommended changes so far.  Some settings like project brief, logo, etc. are left blank for the time being since we have not discussed content for them yet.

> Also how do you handle version numbers? In the NCEPLIBS we have a file at the top level called VERSION which contains the version number. From there it is pulled in by the build system into the docs and the code.

We haven't discussed yet how we'd like to version the doxygen documentation so that's one of the fields we've left empty currently. That sounds like a clean solution though.  I think were still nailing down the basic templates at this point and then we will get to some of the polishing like version numbers shortly after.

edwardhartnett commented:
OK, can you move your Doxyfile to a subdirectory called "docs"?

Then, once you have committed these changes, I can submit a PR which turns the Doxyfile into Doxyfile.in, and causes CMake to generate the Doxyfile.

At that time we can add the VERSION file, and use it to store the software version. That's how the documentation can then be versioned...

MatthewMasarik-NOAA commented:
> OK, can you move your Doxyfile to a subdirectory called "docs"?
> 
> Then, once you have committed these changes, I can submit a PR which turns the Doxyfile into Doxyfile.in, and causes CMake to generate the Doxyfile.
> 
> At that time we can add the VERSION file, and use it to store the software version. That's how the documentation can then be versioned...

Yes.  I'll make those changes after we have finalized our templates.  This PR does not add any of the documentation, it's just to get developer agreement on doxygen headers.

MatthewMasarik-NOAA commented:
A new PR which takes the headers from above, lists them in a markdown file, and adds them in a 'docs' folder has just been submitted, #585 .  The new PR replaces this one.  The new PR has no markup to the code, which is what I wanted for this initial submission.

edwardhartnett commented:
Is this not going to be merged?

MatthewMasarik-NOAA commented:
Hi @edwardhartnett , I closed this branch and added #585 in it's place.  The reason is I had intended to initially just to get consensus on the headers, and not add any marked up code or documentation.  #585 takes into account all the comments that had been made here, and put into the Header reference (./docs/ww3_doxy_tmpl.md).  I also added an issue #586 that we will start incrementally committing marked up code and documentation.  I apologize for making this PR a little more confusing than it needed to be, this was a learning experience for me.  Note too that the issue #586 has the 'docs/' folder you requested, and has the Doxygen.in file inside as well, so if you had wanted to make a PR for changes to it please feel free.

edwardhartnett commented:
OK, great! Good to see all this progress!


On Mon, Jan 24, 2022 at 6:33 AM Matthew Masarik ***@***.***>
wrote:

> Hi @edwardhartnett <https://github.com/edwardhartnett> , I closed this
> branch and added #585 <https://github.com/NOAA-EMC/WW3/pull/585> in it's
> place. The reason is I had intended to initially just to get consensus on
> the headers, and not add any marked up code or documentation. #585
> <https://github.com/NOAA-EMC/WW3/pull/585> takes into account all the
> comments that had been made here, and into the Header reference
> (./docs/ww3_doxy_tmpl.md). I also added an issue #586
> <https://github.com/NOAA-EMC/WW3/issues/586> that we will start
> incrementally committing marked up code and documentation. I apologize for
> making this PR a little more confusing than it needed to be, this was a
> learning experience for me. Note too that the issue #586
> <https://github.com/NOAA-EMC/WW3/issues/586> has the 'docs/' folder you
> requested, and has the Doxygen.in file inside as well, so if you had wanted
> to make a PR for changes to it please feel free.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/WW3/pull/562#issuecomment-1020106283>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJIOMMCH5F4JJOKODL32VX3UXVILDANCNFSM5KHOCYKA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

MatthewMasarik-NOAA commented:
Thank you, @edwardhartnett !
